### ISSUE TITLE
$schema language, "MUST be normalized," is ambiguous

### ISSUE THREAD

**Issue by ssilverman** [json-schema-org/json-schema-spec#899]
[8.1.1. The "$schema" Keyword](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.1.1) states: "this URI MUST be normalized."

This can be read in two ways:
1. MUST be normalized by the validator reading the URI.
2. MUST already be in normalized form.

**Comment by karenetheridge** [comment]
There are a few other places in the spec that say "MUST be normalized" that should also be clarified similarly.

**Comment by handrews** [comment]
"MUST be" means "MUST exist in the state of".  I think it should be clear from the fact that all of these keyword descriptions are talking about usage, not processing.

I realize standards-ese is its own weird dialect only marginally related to English, and I value clarity.  But I'm not aware of this being a problem in the decade or so the specification has existed, and the language feels similar to language in other RFCs.

**Comment by awwright** [comment]
I tend to agree with @handrews: even if in plain english this seems ambiguous, the follow-up question is "MUST be normalized _by who_?" And the safe answer is "by everyone, as early as possible"

**Comment by ssilverman** [comment]
> I tend to agree with @handrews: even if in plain english this seems ambiguous, the follow-up question is "MUST be normalized _by who_?" And the safe answer is "by everyone, as early as possible"

Your use of "by everyone" includes the validator.

**Comment by awwright** [comment]
@ssilverman Sure, but since normalization isn't supposed to affect equality testing, that doesn't mean much. Implementations can do whatever they want so long as the behavior is the same.

I _suppose_ it would be easier to comprehend if we s/be/appear/, like so:

> this URI MUST appear normalized

**Comment by ssilverman** [comment]
Yes, I like being more explicit and removing ambiguity where I can.

**Comment by ssilverman** [comment]
I'll see if I can come up with a concise PR.

**Comment by Relequestual** [comment]
In those two instanced you've highlighted, the spec document is talking about the JSON fields, and what they MUST be.

For example...

> The value of this keyword MUST be a URI (containing a scheme) and this URI MUST be normalized. 

An implementation can't MAKE the value of the keyword be a URI... it can only take what the schema author has given... and throw an error if it's not.

I would assume the same context follows, the schema author, not the implementation.

As such, just as the implementation might throw an error if the value isn't a valid URI, it might also expect the URI to be normalised. The spec certinly doesn't allow the implementation to change the value in such a situation, only place boundaries on what its expectations should be, and when it might want to throw errors.

**Comment by Relequestual** [comment]
I think the intent is for it "to have been normalised" as opposed to "you (or something) should normalise it now"

**Comment by Relequestual** [comment]
`s/be/appear/` seems to be the best solution here.

**Comment by handrews** [comment]
@Relequestual @ssilverman having thought on this more I'm actually fairly strongly against making a change here.

All keywords define their values, and state what the values need to be.  All of them use the same style of language, expressing what schema authors need to provide (an integer, a number,  a boolean, a URI reference, etc.).  None of them discuss coercion by an implementation at all.

Keywords that take a boolean say  `The value of these keywords MUST be a boolean.`

This means what it says:  When you write a schema you have to use a boolean value with that keyword.  It does ***NOT*** mean "the implementation should interpret the value as a boolean," which would basically mean an automatic typecast based on truthiness.

This statement about the normalization of the URI is the same.  I do not want to start messing with the language in a way that implies that ***any*** of this has anything to do with implementations coercing things.

This is standardized language.  @ssilverman I'm sorry you read it differently, but it's worked fine for years and it's very plain English.  It is better for us to use consistent wording across all keywords and make it clear that all of them are talking about what you have to put there as a schema author.

I'm strongly against "appear" in particular.  A URI does not appear to be normalized- it either is or it isn't.

Far from removing ambiguity, "appear" would add it by suggesting that there are URIs that could appear to be normalized without being normalized.

**Comment by handrews** [comment]
Based on the thumbs-up and after checking with @ssilverman on slack, I believe we have resolved this in favor of leaving it as-is.
